File: apps/api/app/schemas/loyalty.py
import uuid
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel, Field


class LoyaltyWalletRead(BaseModel):
    balance_points: int
    lifetime_earned: int
    lifetime_burned: int
    updated_at: datetime

    model_config = {"from_attributes": True}


class LoyaltyLedgerEntryRead(BaseModel):
    id: uuid.UUID
    delta_points: int
    reason: str
    order_id: uuid.UUID | None
    created_at: datetime

    model_config = {"from_attributes": True}


class ReviewCreate(BaseModel):
    product_id: uuid.UUID
    order_item_id: uuid.UUID | None = None
    rating: int = Field(ge=1, le=5)
    title: str | None = Field(default=None, max_length=255)
    body: str | None = None


class ReviewRead(BaseModel):
    id: uuid.UUID
    user_id: uuid.UUID
    product_id: uuid.UUID
    order_item_id: uuid.UUID | None
    rating: int
    title: str | None
    body: str | None
    is_published: bool
    created_at: datetime

    model_config = {"from_attributes": True}


class CouponValidateRead(BaseModel):
    code: str
    coupon_type: str
    value: Decimal
    min_order_pkr: int | None
    ends_at: datetime | None

    model_config = {"from_attributes": True}


class BannerRead(BaseModel):
    id: uuid.UUID
    slot: str
    image_url: str
    link_url: str | None
    title: str | None
    subtitle: str | None
    sort_order: int

    model_config = {"from_attributes": True}


class TestimonialRead(BaseModel):
    id: uuid.UUID
    name: str
    body: str
    rating: int
    image_url: str | None
    is_featured: bool

    model_config = {"from_attributes": True}

File: apps/web/lib/loyalty.ts
import { apiFetch } from "@/lib/api";

export interface LoyaltyWalletRead {
  balance_points: number;
  lifetime_earned: number;
  lifetime_burned: number;
  updated_at: string;
}

export interface LoyaltyLedgerEntry {
  id: string;
  delta_points: number;
  reason: string;
  order_id: string | null;
  created_at: string;
}

export interface ReviewRead {
  id: string;
  user_id: string;
  product_id: string;
  order_item_id: string | null;
  rating: number;
  title: string | null;
  body: string | null;
  is_published: boolean;
  created_at: string;
}

export interface ReviewCreate {
  product_id: string;
  order_item_id?: string | null;
  rating: number;
  title?: string | null;
  body?: string | null;
}

export interface CouponValidateRead {
  code: string;
  coupon_type: "percent" | "fixed" | "free_shipping";
  value: number;
  min_order_pkr: number | null;
  ends_at: string | null;
}

export interface BannerRead {
  id: string;
  slot: string;
  image_url: string;
  link_url: string | null;
  title: string | null;
  subtitle: string | null;
  sort_order: number;
}

export interface TestimonialRead {
  id: string;
  name: string;
  body: string;
  rating: number;
  image_url: string | null;
  is_featured: boolean;
}

export const loyaltyApi = {
  getWallet: (token: string) =>
    apiFetch<LoyaltyWalletRead>("/loyalty/me", { token }),

  getLedger: (token: string, limit = 50) =>
    apiFetch<LoyaltyLedgerEntry[]>(`/loyalty/me/ledger?limit=${limit}`, { token }),

  submitReview: (body: ReviewCreate, token: string) =>
    apiFetch<ReviewRead>("/reviews", { method: "POST", body, token }),

  listReviews: (productId: string) =>
    apiFetch<ReviewRead[]>(`/reviews?product_id=${productId}`),

  validateCoupon: (code: string) =>
    apiFetch<CouponValidateRead>(`/coupons/${encodeURIComponent(code)}/validate`),

  listBanners: (slot?: string) =>
    apiFetch<BannerRead[]>(slot ? `/banners?slot=${slot}` : "/banners"),

  listTestimonials: () => apiFetch<TestimonialRead[]>("/testimonials"),
};
